codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -579,102 +318,85 @@ public boolean isEmpty() {
     }
 
     @Override
-    public Headers<T> add(Headers<T> headers) {
+    public Headers<T> add(Headers<? extends T> headers) {
         checkNotNull(headers, "headers");
-
-        add0(headers);
+        if (headers == this) {
+            throw new IllegalArgumentException("can't add to itself.");
+        }
+        for (Entry<? extends T, ? extends T> header : headers) {
+            add(header.getKey(), header.getValue());
+        }
         return this;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public Headers<T> set(T name, T value) {
-        name = convertName(name);
+        validateName(name);
         checkNotNull(value, "value");
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        remove0(h, i, name);
-        add0(h, i, name, value);
+        Object oldValue = map.put(name, value);
+        updateSizeAfterSet(oldValue, 1);
         return this;
     }
 
     @Override
     public Headers<T> set(T name, Iterable<? extends T> values) {
-        name = convertName(name);
+        validateName(name);
         checkNotNull(values, "values");
-
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        remove0(h, i, name);
-        for (T v : values) {
-            if (v == null) {
-                break;
-            }
-            add0(h, i, name, v);
+        List<T> list = newList();
+        for (T value : values) {
+            list.add(value);
```

Comment:
We check `name` and `values` but not each `value` that we are adding to the list.